src/clojure/nrepl/middleware/interruptible_eval.clj

```diff
@@ -47,88 +47,65 @@
-           :init #(do (set! *compile-path* (@bindings #'*compile-path*))
-                      (set! *1 (@bindings #'*1))
-                      (set! *2 (@bindings #'*2))
-                      (set! *3 (@bindings #'*3))
-                      (set! *e (@bindings #'*e))
-                      (when-some [v (resolve '*print-namespace-maps*)]
-                        (var-set v (@bindings v))))
-           :read (if (string? code)
-                   (let [reader (source-logging-pushback-reader code line column)]
-                     #(read {:read-cond :allow :eof %2} reader))
-                   (let [code (.iterator ^Iterable code)]
-                     #(or (and (.hasNext code) (.next code)) %2)))
-           :prompt (fn [])
-           :need-prompt (constantly false)
-           ;; TODO: pretty-print?
-           :print (fn [v]
-                    (reset! bindings (assoc (capture-thread-bindings)
-                                            #'*3 *2
-                                            #'*2 *1
-                                            #'*1 v))
-                    (.flush ^Writer err)
-                    (.flush ^Writer out)
-                    (reset! session (maybe-restore-original-ns @bindings))
-                    (t/send transport (response-for msg
-                                                    {:value v
-                                                     :ns (-> *ns* ns-name str)})))
-           ;; TODO: customizable exception prints
-           :caught (fn [e]
-                     (let [root-ex (#'clojure.main/root-cause e)
-                           previous-cause (.getCause e)]
-                       ;; Check if the root cause or previous cause of the exception
-                       ;; is a ThreadDeath exception. In case the exception is a
-                       ;; CompilerException, the root cause is not returned by
-                       ;; the root-cause function, so we check the previous cause
-                       ;; instead.
-                       (when-not (or (instance? ThreadDeath root-ex)
-                                     (instance? ThreadDeath previous-cause))
-                         (reset! bindings (assoc (capture-thread-bindings) #'*e e))
-                         (reset! session (maybe-restore-original-ns @bindings))
-                         (t/send transport (response-for msg {:status :eval-error
-                                                              :ex (-> e class str)
-                                                              :root-ex (-> root-ex class str)}))
-                         (clojure.main/repl-caught e)))))
-          (finally
-            (.flush ^Writer out)
-            (.flush ^Writer err)))))
-    (maybe-restore-original-ns @bindings)))
+      (try
+        (clojure.main/repl
+         :eval (if eval (find-var (symbol eval)) clojure.core/eval)
+         ;; clojure.main/repl paves over certain vars even if they're already thread-bound
```

Comment:
Btw, do you know what exactly is paved over by cloure.main/repl?

Reply:
I should have removed this comment, since it's now irrelevant. Look here for the list:
See here https://github.com/clojure/clojure/blob/master/src/clj/clojure/main.clj#L81

It's irrelevant because of the get/pop/push sequence which overwrites them with session values.

So if Clojure 1.11 adds a new dynvar we get the default value on session creation (dummy evaluate) and after that we force the session value in init, so we are in control.

Reply:
Got it. Thanks for the explanation!